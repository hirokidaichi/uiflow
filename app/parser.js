var SECTION = /\[([^#\]]*)(#+)?\]/;
var END_OF_SEES = /^-+$/;
var END_OF_ACTION = /=+(?:{([^}]+)})?=+>\s*([^:\]]*)/;
var WHITE_LINE = /^\s*$/;

var lexer = module.exports.lexer = function(text) {
    var state = {};
    return text.split(/\n/).map(parseByLine);
};
var parseError = function(fileName) {
    return function(message, lineNumber) {
        var error = new Error([
            fileName || "(anon)", ":", lineNumber, ": error:", message
        ].join(""));
        error.lineNumber = lineNumber;
        error.fileName = fileName;
        throw error;
    };
};

var parseByLine = function(line, num) {
    if (SECTION.test(line)) {
        var rank = RegExp.$2.length;
        return ["section", RegExp.$1, rank, num];
    }
    if (END_OF_SEES.test(line)) {
        return ["endofsee", num];
    }
    if (END_OF_ACTION.test(line)) {
        return ["endofaction", RegExp.$2, RegExp.$1, num];
    }
    if (WHITE_LINE.test(line)) {
        return ["whiteline", num];
    }
    return ["text", line, num];
};

var parseTags = function(listOfNode, fileName) {
    var tree = {};
    var nId = 1;
    var currentSection = null;
    var lastAction;
    var actions;
    var errorMessage = parseError(fileName);
    listOfNode.forEach(function(node) {
        var tag = node[0];
        if (tag == "whiteline") {
            return;
        }
        if (tag == "section") {
            currentSection = node[1];
            if (lastAction && !lastAction.direction) {
                lastAction.direction = currentSection;
            }

            if (tree[currentSection]) {
                errorMessage("Duplicated section:" + currentSection, node[3]);
            }
            tree[currentSection] = {
                name: currentSection,
                rank: node[2],
                lines: node[3],
                see: [],
                id: nId++,
                actions: [{
                    text: [],
                    direction: null
                }],
                state: "see"
            };
        }
        if (tag == "endofsee") {
            if (!tree[currentSection]) {
                errorMessage("Undefined section [" + currentSection + "]", node[1]);
            }
            if (tree[currentSection].state == "action") {
                errorMessage("Duplicated sees" + "\tL:", node[1]);
            }
            tree[currentSection].state = "action";
        }
        if (tag == "endofaction") {
            if (!tree[currentSection]) {
                errorMessage("Undefined section" + "\tL:", node[3]);
            }
            actions = tree[currentSection].actions;
            actions[actions.length - 1].direction = node[1];
            actions[actions.length - 1].edge = node[2];

            tree[currentSection].state = "endaction";

        }
        if (tag == "text") {
            if (!tree[currentSection]) {
                errorMessage("Undefined section" + "\tL:", node[2]);
            }
            var state = tree[currentSection].state;
            if (state == "see") {
                tree[currentSection].see.push(node[1]);
            }
            if (state == "endaction") {
                tree[currentSection].actions.push({
                    text: [],
                    direction: null
                });
                state = tree[currentSection].state = "action";
            }
            if (state == "action") {
                actions = tree[currentSection].actions;
                actions[actions.length - 1].text.push(node[1]);
                lastAction = actions[actions.length - 1];
            }
        }
    });
    return tree;
};
var parse = module.exports.parse = function(text, fileName) {
    return parseTags(lexer(text), fileName);
};
