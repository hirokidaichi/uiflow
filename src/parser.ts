import { Action, Tag } from './interfaces'

const SECTION = /\[([^#\]]*)(#+)?]/
const END_OF_SEES = /^-+$/
const END_OF_ACTION = /=+(?:{([^}]+)})?=+>\s*([^:\]]*)/
const WHITE_LINE = /^\s*$/

interface LineParseResult {
  tag: Tag
  num: number
  rank?: number
  actionText?: string
  api?: string

  title?: string
  text?: string
}

type ParseResult = LineParseResult[]

type State = 'action' | 'see' | 'endaction'

interface TreeNode {
  id: number
  name: string
  rank: number
  lines: number
  see: string[]
  actions: Action[]
  state: State
}

type Tree = Record<string, TreeNode>

export class ParseError extends Error {
  public constructor(
    message: string,
    public lineNumber: number,
    public fileName: string
  ) {
    super(message)
  }
}

const lexer = (text: string): ParseResult => {
  return text.split(/\n/).map(parseByLine)
}
const parseError = (fileName: string) => {
  return (message: string, lineNumber: number) => {
    const error = new ParseError(
      [
        fileName !== null || '(anon)',
        ':',
        lineNumber,
        ': error:',
        message,
      ].join(''),
      lineNumber,
      fileName
    )
    throw error
  }
}

const parseByLine = (line: string, num: number): LineParseResult => {
  const section = line.match(SECTION)
  if (section !== null) {
    const rank = section[2].length
    const title = section[1]
    return {
      tag: 'section',
      title,
      rank,
      num,
    }
  }

  const endOfSees = line.match(END_OF_SEES)
  if (endOfSees !== null) {
    return {
      tag: 'endOfSee',
      num,
    }
  }

  const endOfAction = line.match(END_OF_ACTION)
  if (endOfAction !== null) {
    const actionText = endOfAction[2]
    const api = endOfAction[1]
    return {
      tag: 'endOfAction',
      actionText,
      api,
      num,
    }
  }

  const whiteLine = line.match(WHITE_LINE)
  if (whiteLine !== null) {
    return {
      tag: 'whiteLine',
      num,
    }
  }
  return { tag: 'text', num: num }
}

const parseTags = (listOfNode: LineParseResult[], fileName: string): Tree => {
  const tree: Tree = {}
  let nId = 1
  let currentSection: string
  let lastAction: Action
  let actions: Action[]
  const errorMessage = parseError(fileName)

  listOfNode.forEach((node) => {
    const tag = node.tag
    if (tag === 'whiteLine') {
      return
    }
    if (tag === 'section') {
      currentSection = node.title ?? ''
      if (lastAction?.direction == null) {
        lastAction.direction = currentSection
      }

      if (currentSection in tree) {
        errorMessage('Duplicated section:' + currentSection, node.num)
      }

      tree[currentSection] = {
        name: currentSection,
        rank: node.rank ?? 0,
        lines: node.num ?? 0,
        see: [],
        id: nId++,
        actions: [
          {
            text: [],
            direction: null,
          },
        ],
        state: 'see',
      }
    }

    if (tag === 'endOfSee') {
      if (!(currentSection in tree)) {
        errorMessage(`Undefined section ${currentSection}]`, node.num)
      }

      if (tree[currentSection].state === 'action') {
        errorMessage('Duplicated sees' + '\tL:', node.num)
      }
      tree[currentSection].state = 'action'
    }

    if (tag === 'endOfAction') {
      if (!(currentSection in tree)) {
        errorMessage('Undefined section' + '\tL:', node.num)
      }

      let state = tree[currentSection].state

      if (state === 'see') {
        tree[currentSection].see.push(node.text ?? '')
      }

      if (state === 'endaction') {
        tree[currentSection].actions.push({
          text: [],
          direction: null,
        })
        state = tree[currentSection].state = 'action'
      }

      if (state === 'action') {
        actions = tree[currentSection].actions
        actions[actions.length - 1].text.push(node.actionText ?? '')
      }
    }
  })

  return tree
}

export function parse(text: string, fileName: string): Tree {
  return parseTags(lexer(text), fileName)
}
