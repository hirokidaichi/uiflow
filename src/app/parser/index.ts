import { parse } from './parser'

export const parser = {
  parse,
} as const

export default parser
