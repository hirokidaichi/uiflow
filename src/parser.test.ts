import { parse } from './parser'
test('parser', () => {
  parse(`[hoge]`, 'hoge.uif')
})
