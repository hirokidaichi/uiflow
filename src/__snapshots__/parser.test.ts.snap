// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser 1`] = `
{
  "はじめよう": {
    "actions": [
      {
        "direction": "",
        "edge": "はじめよう",
        "text": [
          "",
        ],
      },
      {
        "direction": "非ログインメイン画面",
        "text": [
          "",
        ],
      },
    ],
    "id": 3,
    "lines": 13,
    "name": "はじめよう",
    "rank": 0,
    "see": [
      "ヒント情報",
      "はじめるボタン",
    ],
    "state": "action",
  },
  "ランキング": {
    "actions": [
      {
        "direction": null,
        "edge": null,
        "text": [],
      },
    ],
    "id": 7,
    "lines": 48,
    "name": "ランキング",
    "rank": 0,
    "see": [
      "ランキング詳細",
    ],
    "state": "see",
  },
  "分野": {
    "actions": [
      {
        "direction": null,
        "edge": null,
        "text": [],
      },
    ],
    "id": 6,
    "lines": 45,
    "name": "分野",
    "rank": 0,
    "see": [
      "分野詳細",
    ],
    "state": "see",
  },
  "初回通知": {
    "actions": [
      {
        "direction": "国選択",
        "edge": null,
        "text": [
          "",
        ],
      },
    ],
    "id": 1,
    "lines": 1,
    "name": "初回通知",
    "rank": 0,
    "see": [
      "通知の確認ダイアログ",
    ],
    "state": "action",
  },
  "利用規約": {
    "actions": [
      {
        "direction": null,
        "edge": null,
        "text": [],
      },
    ],
    "id": 5,
    "lines": 42,
    "name": "利用規約",
    "rank": 0,
    "see": [
      "利用規約の表示",
    ],
    "state": "see",
  },
  "国選択": {
    "actions": [
      {
        "direction": "はじめよう",
        "edge": null,
        "text": [
          "",
        ],
      },
    ],
    "id": 2,
    "lines": 6,
    "name": "国選択",
    "rank": 0,
    "see": [
      "リージョンの選択",
      "デフォルトで日本",
      "OKボタン",
    ],
    "state": "action",
  },
  "非ログインメイン画面": {
    "actions": [
      {
        "direction": "",
        "edge": "利用規約",
        "text": [
          "",
        ],
      },
      {
        "direction": "",
        "edge": "分野",
        "text": [
          "",
        ],
      },
      {
        "direction": "",
        "edge": "ログイン確認",
        "text": [
          "",
        ],
      },
      {
        "direction": "",
        "edge": "ランキング",
        "text": [
          "",
        ],
      },
    ],
    "id": 4,
    "lines": 21,
    "name": "非ログインメイン画面",
    "rank": 0,
    "see": [
      "ログインすると全ての機能を使えますアラート",
      "フィード",
      "通知アイコン",
      "検索アイコン",
      "ホーム",
      "分野",
      "Qボタン",
      "ランキング",
      "マイページ(N)",
    ],
    "state": "endaction",
  },
}
`;
